perfkitbenchmarker/linux_virtual_machine.py

```diff
@@ -954,3 +956,191 @@ def MoveFile(self, target, source_path, remote_path=''):
 
     # Copies the file to its final destination in the container
     target.ContainerCopy(file_name, remote_path)
+
+
+class JujuMixin(DebianMixin):
+    """
+    Class to allow running juju-deployed workloads
+
+    Bootstraps a Juju environment using the manual provider:
+    https://jujucharms.com/docs/stable/config-manual
+    """
+
+    OS_TYPE = 'juju'
+
+    isController = False
+
+    """
+    A reference to the juju controller, useful when operations occur against
+    a unit's VM but need to be preformed from the controller.
+    """
+    controller = None
+
+    units = []
+
+    environments_yaml = """
+    default: perfkit
+
+    environments:
+        perfkit:
+            type: manual
+            bootstrap-host: {0}
+    """
+
+    def juju_bootstrap(self):
+        resp, _ = self.RemoteHostCommand("juju bootstrap")
+        pass
+
+    def juju_addmachine(self, ip):
+        resp, _ = self.RemoteHostCommand("juju add-machine ssh:%s" % ip)
+
+    def juju_configure_environment(self):
+        if self.isController:
+
+            resp, _ = self.RemoteHostCommand("mkdir -p ~/.juju")
+
+            f = tempfile.NamedTemporaryFile(delete=False)
+            f.write(self.environments_yaml.format(self.internal_ip))
+            f.close()
+            self.RemoteCopy(f.name, "~/.juju/environments.yaml")
+            os.unlink(f.name)
+        pass
+
+    def juju_environment(self, environment=''):
+        # TODO: set environment on demand
+        try:
+            output, _ = self.RemoteHostCommand('juju switch')
+            return output.strip()
+        except:
+            pass
+        return None
+
+    def juju_run(self, cmd):
+        try:
+            output, _ = self.RemoteHostCommand(cmd)
+            return output.strip()
+        except:
+            pass
+        return None
+
+    def juju_status(self):
+        """
+        Return the status of the Juju environment.
+        """
+        try:
+
+            output, _ = self.RemoteHostCommand('juju status --format=json')
+            return output.strip()
+        except:
+            pass
+        return None
+
+    def juju_version(self):
+        try:
+            output, _ = self.RemoteHostCommand('juju version')
+            return output.strip()
+        except:
+            pass
+        return None
+
+    def Set(self, service, params=[]):
+        ' '.join(params)
+        try:
+            output, _ = self.RemoteHostCommand(
+                'juju set %s %s' % (service, ' '.join(params)))
+            return output.strip()
+        except:
+            pass
+
+
+    def Wait(self, timeout=900):
```

Comment:
nit: for methods that are going to be called elsewhere could you try working juju into the name so that it's obvious where the method comes from? vm.Wait() looks very generic whereas vm.JujuWait() (or similar) would make things more clear